MapStruct does not always use builders in target-object

Hi,

we try to use MapStruct with Lombok builders but MapStruct does **not always** seem to recognize and use the builders. Our particular problematic case is when the **target** object has a field which only has a builder.

Used MapStruct version: 1.3.1.FINAL
Used Lombok version: 1.18.10

It would be very helpful for us if you could let us know if this is considered a bug or not.
Looking forward to your feedback!

**Here is an example:**

Source POJO:
```
package example.model.source;

import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class Car {
	private String model;
}
```

Target POJOs:
```
package example.model.target;

import lombok.Builder;
import lombok.Data;

@Data
@Builder
public class CarInsurance {
	private CarDetail carDetail;   // this is the problematic case
}

@Data
@Builder
public class CarDetail {
	private String model;
}
```

Mapper:
```
package example.mapper;

import example.model.source.Car;
import example.model.target.CarInsurance;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;

@Mapper
public interface CarInsuranceMapper {
	@Mapping(source = "model", target = "carDetail.model")
	void map(Car source, @MappingTarget CarInsurance target);
}
```

MapStruct generates this mapper:
```
package example.mapper;

import example.model.source.Car;
import example.model.target.CarDetail;
import example.model.target.CarInsurance;
import javax.annotation.Generated;
import org.springframework.stereotype.Component;

@Generated(
    value = "org.mapstruct.ap.MappingProcessor"
)
@Component
public class CarInsuranceMapperImpl implements CarInsuranceMapper {

    @Override
    public void map(Car source, CarInsurance target) {
        if ( source == null ) {
            return;
        }

        if ( target.getCarDetail() == null ) {
            target.setCarDetail( new CarDetail() );  // THIS CREATES A COMPILE ERROR
        }
        carToCarDetail( source, target.getCarDetail() );
    }

    protected void carToCarDetail(Car car, CarDetail mappingTarget) {
        if ( car == null ) {
            return;
        }

        mappingTarget.setModel( car.getModel() );
    }
}
```

And the compiler fails with:
```
CarInsuranceMapperImpl.java:22: error: constructor CarDetail in class CarDetail cannot be applied to given types;
            target.setCarDetail( new CarDetail() );

```

It would work if MapStruct would generate:
```
if ( target.getCarDetail() == null ) {
    target.setCarDetail( CarDetail.builder().build() );
}
```

The annotation processor ordering in the Gradle build:
```
	implementation 'org.mapstruct:mapstruct'
	implementation 'org.projectlombok:lombok'
	annotationProcessor 'org.mapstruct:mapstruct-processor'
	annotationProcessor 'org.projectlombok:lombok'
```

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/MethodReference.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/MethodReference.java
@@ -6,6 +6,7 @@
 package org.mapstruct.ap.internal.model;
 
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
@@ -58,8 +59,10 @@ public class MethodReference extends ModelElement implements Assignment {
     private final Type definingType;
     private final List<ParameterBinding> parameterBindings;
     private final Parameter providingParameter;
+    private final List<MethodReference> methodsToChain;
     private final boolean isStatic;
     private final boolean isConstructor;
+    private final boolean isMethodChaining;
 
     /**
      * Creates a new reference to the given method.
@@ -95,6 +98,8 @@ public class MethodReference extends ModelElement implements Assignment {
         this.isStatic = method.isStatic();
         this.name = method.getName();
         this.isConstructor = false;
+        this.methodsToChain = Collections.emptyList();
+        this.isMethodChaining = false;
    }
 
     private MethodReference(BuiltInMethod method, ConversionContext contextParam) {
@@ -111,6 +116,8 @@ public class MethodReference extends ModelElement implements Assignment {
         this.isStatic = method.isStatic();
         this.name = method.getName();
         this.isConstructor = false;
+        this.methodsToChain = Collections.emptyList();
+        this.isMethodChaining = false;
     }
 
     private MethodReference(String name, Type definingType, boolean isStatic) {
@@ -127,6 +134,8 @@ public class MethodReference extends ModelElement implements Assignment {
         this.providingParameter = null;
         this.isStatic = isStatic;
         this.isConstructor = false;
+        this.methodsToChain = Collections.emptyList();
+        this.isMethodChaining = false;
     }
 
     private MethodReference(Type definingType, List<ParameterBinding> parameterBindings) {
@@ -142,6 +151,8 @@ public class MethodReference extends ModelElement implements Assignment {
         this.providingParameter = null;
         this.isStatic = false;
         this.isConstructor = true;
+        this.methodsToChain = Collections.emptyList();
+        this.isMethodChaining = false;
 
         if ( parameterBindings.isEmpty() ) {
             this.importTypes = Collections.emptySet();
@@ -157,6 +168,24 @@ public class MethodReference extends ModelElement implements Assignment {
 
             this.importTypes = Collections.unmodifiableSet( imported );
         }
+    }
+
+    private MethodReference(MethodReference... references) {
+        this.name = null;
+        this.definingType = null;
+        this.sourceParameters = Collections.emptyList();
+        this.returnType = null;
+        this.declaringMapper = null;
+        this.importTypes = Collections.emptySet();
+        this.thrownTypes = Collections.emptyList();
+        this.isUpdateMethod = false;
+        this.contextParam = null;
+        this.parameterBindings = null;
+        this.providingParameter = null;
+        this.isStatic = false;
+        this.isConstructor = false;
+        this.methodsToChain = Arrays.asList( references );
+        this.isMethodChaining = true;
     }
 
     public MapperReference getDeclaringMapper() {
@@ -267,6 +296,12 @@ public class MethodReference extends ModelElement implements Assignment {
         if ( isStatic() ) {
             imported.add( definingType );
         }
+        if ( isMethodChaining() ) {
+            for ( MethodReference methodToChain : methodsToChain ) {
+                imported.addAll( methodToChain.getImportTypes() );
+            }
+        }
+
         return imported;
     }
 
@@ -275,6 +310,12 @@ public class MethodReference extends ModelElement implements Assignment {
         List<Type> exceptions = new ArrayList<>( thrownTypes );
         if ( assignment != null ) {
             exceptions.addAll( assignment.getThrownTypes() );
+        }
+        if ( isMethodChaining() ) {
+            for ( MethodReference methodToChain : methodsToChain ) {
+                exceptions.addAll( methodToChain.getThrownTypes() );
+            }
+
         }
         return exceptions;
     }
@@ -309,6 +350,14 @@ public class MethodReference extends ModelElement implements Assignment {
 
     public boolean isConstructor() {
         return isConstructor;
+    }
+
+    public boolean isMethodChaining() {
+        return isMethodChaining;
+    }
+
+    public List<MethodReference> getMethodsToChain() {
+        return methodsToChain;
     }
 
     public List<ParameterBinding> getParameterBindings() {
@@ -383,6 +432,10 @@ public class MethodReference extends ModelElement implements Assignment {
 
     public static MethodReference forConstructorInvocation(Type type, List<ParameterBinding> parameterBindings) {
         return new MethodReference( type, parameterBindings );
+    }
+
+    public static MethodReference forMethodChaining(MethodReference... references) {
+        return new MethodReference( references );
     }
 
     @Override

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/ObjectFactoryMethodResolver.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/ObjectFactoryMethodResolver.java
@@ -139,6 +139,10 @@ public class ObjectFactoryMethodResolver {
     }
 
     public static MethodReference getBuilderFactoryMethod(Method method, BuilderType builder ) {
+        return getBuilderFactoryMethod( method.getReturnType(), builder );
+    }
+
+    public static MethodReference getBuilderFactoryMethod(Type typeToBuild, BuilderType builder ) {
         if ( builder == null ) {
             return null;
         }
@@ -149,7 +153,7 @@ public class ObjectFactoryMethodResolver {
             return null;
         }
 
-        if ( !builder.getBuildingType().isAssignableTo( method.getReturnType() ) ) {
+        if ( !builder.getBuildingType().isAssignableTo( typeToBuild ) ) {
             //TODO print error message
             return null;
         }

--- a/processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java
+++ b/processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java
@@ -419,6 +419,28 @@ public class PropertyMapping extends ModelElement {
 
                 Assignment factory = ObjectFactoryMethodResolver
                     .getFactoryMethod( method, targetType, SelectionParameters.forSourceRHS( rightHandSide ), ctx );
+
+                if ( factory == null && targetBuilderType != null) {
+                    // If there is no dedicated factory method and the target has a builder we will try to use that
+                    MethodReference builderFactoryMethod = ObjectFactoryMethodResolver.getBuilderFactoryMethod(
+                        targetType,
+                        targetBuilderType
+                    );
+
+                    if ( builderFactoryMethod != null ) {
+
+                        MethodReference finisherMethod = BuilderFinisherMethodResolver.getBuilderFinisherMethod(
+                            method,
+                            targetBuilderType,
+                            ctx
+                        );
+
+                        if ( finisherMethod != null ) {
+                            factory = MethodReference.forMethodChaining( builderFactoryMethod, finisherMethod );
+                        }
+                    }
+
+                }
                 return new UpdateWrapper(
                     rhs,
                     method.getThrownTypes(),

--- a/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReference.ftl
+++ b/processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReference.ftl
@@ -18,6 +18,8 @@
         <@includeModel object=definingType/>.<@methodCall/>
     <#elseif constructor>
         new <@includeModel object=definingType/><#if (parameterBindings?size > 0)>( <@arguments/> )<#else>()</#if>
+    <#elseif methodChaining>
+        <#list methodsToChain as methodToChain><@includeModel object=methodToChain /><#if methodToChain_has_next>.</#if></#list>
     <#else>
         <@methodCall/>
     </#if>

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/Car.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/Car.java
@@ -1,0 +1,44 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1997;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class Car {
+    private String model;
+
+    private Car(Builder builder) {
+        this.model = builder.model;
+    }
+
+    public String getModel() {
+        return model;
+    }
+
+    public void setModel(String model) {
+        this.model = model;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private String model;
+
+        public Builder model(String model) {
+            this.model = model;
+            return this;
+        }
+
+        public Car build() {
+            return new Car( this );
+        }
+
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarDetail.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarDetail.java
@@ -1,0 +1,44 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1997;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class CarDetail {
+    private String model;
+
+    private CarDetail(Builder builder) {
+        this.model = builder.model;
+    }
+
+    public String getModel() {
+        return model;
+    }
+
+    public void setModel(String model) {
+        this.model = model;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private String model;
+
+        public Builder model(String model) {
+            this.model = model;
+            return this;
+        }
+
+        public CarDetail build() {
+            return new CarDetail( this );
+        }
+
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarInsurance.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarInsurance.java
@@ -1,0 +1,44 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1997;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class CarInsurance {
+    private CarDetail detail;
+
+    private CarInsurance(Builder builder) {
+        this.detail = builder.detail;
+    }
+
+    public CarDetail getDetail() {
+        return detail;
+    }
+
+    public void setDetail(CarDetail detail) {
+        this.detail = detail;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private CarDetail detail;
+
+        public Builder detail(CarDetail detail) {
+            this.detail = detail;
+            return this;
+        }
+
+        public CarInsurance build() {
+            return new CarInsurance( this );
+        }
+
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarInsuranceMapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/CarInsuranceMapper.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1997;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.MappingTarget;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper
+public interface CarInsuranceMapper {
+
+    CarInsuranceMapper INSTANCE = Mappers.getMapper( CarInsuranceMapper.class );
+
+    @Mapping(source = "model", target = "detail.model")
+    void update(Car source, @MappingTarget CarInsurance target);
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/Issue1997Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1997/Issue1997Test.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1997;
+
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@WithClasses({
+    Car.class,
+    CarDetail.class,
+    CarInsurance.class,
+    CarInsuranceMapper.class
+})
+class Issue1997Test {
+
+    @ProcessorTest
+    void shouldCorrectCreateIntermediateObjectsWithBuilder() {
+        Car source = Car.builder().model( "Model S" ).build();
+        CarInsurance target = CarInsurance.builder().build();
+        assertThat( target.getDetail() ).isNull();
+
+        CarInsuranceMapper.INSTANCE.update( source, target );
+
+        assertThat( target.getDetail() ).isNotNull();
+        assertThat( target.getDetail().getModel() ).isEqualTo( "Model S" );
+    }
+}

